README.md
# OpenSea NFT API Python 3 wrapper
This an API wrapper library for the [OpenSea API](https://docs.opensea.io/reference/api-overview) written in Python 3.

The library provides a simplified interface to fetch a diverse set of NFT data points from OpenSea. 

## Supported endpoints
The wrapper covers the following OpenSea API endpoints:

* Single asset ([/asset](https://docs.opensea.io/reference/retrieving-a-single-asset))
* Single asset contract ([/asset_contract](https://docs.opensea.io/reference/retrieving-a-single-contract))
* Single collection ([/collection](https://docs.opensea.io/reference/retrieving-a-single-collection))
* Collection stats ([/collection/{slug}/stats](https://docs.opensea.io/reference/retrieving-collection-stats))
* Multiple assets] ([/assets](https://docs.opensea.io/reference/getting-assets))
* Multiple collections ([/collections](https://docs.opensea.io/reference/retrieving-collections))
* Multiple events ([/events](https://docs.opensea.io/reference/retrieving-asset-events))
* Multiple bundles ([/bundles](https://docs.opensea.io/reference/retrieving-bundles))
* Asset listings ([/asset/{asset_contract_address}/{token_id}/listings](https://docs.opensea.io/reference/asset-listings))
* Asset offers ([/asset/{asset_contract_address}/{token_id}/offers](https://docs.opensea.io/reference/asset-offers))


## Prerequisite
You need to have an **API key** to use the OpenSea API, and thus
you need one to use this wrapper too. [You can request a key here.](https://docs.opensea.io/reference/request-an-api-key) <br><br>
NOTE: The API key can take over 4-7 days to be delivered. It also requires you to show the project you are working on. 

## Installation
Install with pip:
```bash
virtualenv env && source env/bin/activate
pip install opensea-api
```

### Upgrade
```bash
pip install opensea-api -U
```

## Usage examples

```python
# import the OpenseaAPI object from the opensea module
from opensea import OpenseaAPI

# create an object to interact with the Opensea API (need an api key)
api = OpenseaAPI(apikey="YOUR APIKEY")

# fetch a single asset
contract_address = "0x495f947276749Ce646f68AC8c248420045cb7b5e"
token_id = "66406747123743156841746366950152533278033835913591691491127082341586364792833"
result = api.asset(asset_contract_address=contract_address, token_id=token_id)

# fetch multiple assets
result = api.assets(owner="0xce90a7949bb78892f159f428d0dc23a8e3584d75", limit=3)

# fetch a single contract
result = api.contract(asset_contract_address="0x495f947276749Ce646f68AC8c248420045cb7b5e")

# fetch a single collection
result = api.collection(collection_slug="cryptopunks")

# fetch multiple collections
result = api.collections(asset_owner="0xce90a7949bb78892f159f428d0dc23a8e3584d75", limit=3)

# fetch collection stats
result = api.collection_stats(collection_slug="cryptopunks")

# fetch multiple events
from opensea import utils as opensea_utils

period_end = opensea_utils.datetime_utc(2021, 11, 6, 14, 30)
result = api.events(
    occurred_before=period_end,
    limit=10,
    export_file_name="events.json",
)

# fetch multiple bundles
result = api.bundles(limit=2)

# paginate the events endpoint (cursors are handled internally)
from datetime import datetime, timezone

start_at = datetime(2021, 10, 5, 3, 25, tzinfo=timezone.utc)
finish_at = datetime(2021, 10, 5, 3, 20, tzinfo=timezone.utc)

event_generator = api.events_backfill(start=start_at,
                                      until=finish_at,
                                      event_type="successful")
for event in event_generator:
    if event is not None:
        print(event) # or do other things with the event data
```

[Here's a demo video showcasing the basics.](https://www.youtube.com/watch?v=ga4hTqNRjfw)

## Documentation
* [Wrapper documentation](https://opensea-api.attilatoth.dev)
* [OpenSea API documentation](https://docs.opensea.io/reference/api-overview)



HISTORY.md
# History

## 0.1.8 (2022-xx-xx)
* Add `include_orders` in the assets api 
* Add cursor-based pagination in `assets` endpoint

## 0.1.7 (2022-03-26)
* Add support for [asset listings](https://docs.opensea.io/reference/asset-listings)
    and [asset offers](https://docs.opensea.io/reference/asset-offers) endpoints
* Add `occured_after` and `collection_editor` arguments to events endpoint
* Handle SSL error when making requests
* Docs: add example to paginate the events endpoint (using `events_backfill()`)

## 0.1.6 (2022-02-25)
* Fix /events endpoint pagination (`events_backfill()` function) by
passing only *the cursor hash* and not the full URL to the next request.

## 0.1.5 (2022-02-17)

* Ability to override base_url with any other URL
* Support for cursor-based pagination for /events endpoint (and removed deprecated arguments)
* New function to help paginate the /events endpoint
* Introducing a temporary function to fix the `next` url problem until OpenSea addresses this issue
* Minor docs updates and cleanup


## 0.1.3 (2021-12-03)

* Ability to reach all endpoints from one `OpenseaAPI` object
* API key support (Opensea requires it from now on)

## 0.1.0 (2021-11-07)

* First release on PyPI.


tests/__init__.py
"""Unit test package for opensea."""


opensea/utils.py
from datetime import datetime, timezone


def export_file(content, file_name):
    """Creates a new file with the specified content and file name. If the file
    already exists, overwrites it.

    Args:
        content (str or bytes): Content to be inserted into the file.
        file_name (str): Name of the file to be created. Eg. 'export.json'.
    """
    with open(file_name, "wb") as f:
        f.write(content)


def str_to_datetime_utc(str):
    """Converts a string into UTC datetime object.

    Args:
        str (str): String timestamp.

    Returns:
        datetime: Datetime object.
    """
    return datetime.fromisoformat(str).replace(tzinfo=timezone.utc)


def datetime_utc(year, month, day, hour, minute):
    """Returns a datetime object in UTC timezone

    Args:
        year (int): eg. 2021
        month (int): between 1-12
        day (int): between 1-7
        hour (int): between 0-23
        minute (int): between 0-59

    Returns:
        datetime
    """
    return datetime(year, month, day, hour, minute, tzinfo=timezone.utc)


opensea/opensea.py
import requests
from datetime import datetime
from opensea import utils


class OpenseaBase:

    def __init__(self, endpoint, version="v1",
                 base_url="https://api.opensea.io/api"):
        """Base class to interact with the OpenSea API and fetch NFT data.

        Args:
            endpoint (str): OpenSea API endpoint, eg. 'asset' or 'collections'
            version (str, optional): API version. Defaults to "v1"
            base
        """
        self.api_url = f"{base_url}/{version}/{endpoint}"

    def _make_request(self, params=None, export_file_name="",
                      return_response=False):
        """Makes a request to the OpenSea API and returns either a response
        object or dictionary.

        Args:
            params (dict, optional): Query parameters to include in the
            request. Defaults to None.
            export_file_name (str, optional): In case you want to download the
            data into a file,
            specify the filename here. Eg. 'export.json'. Be default, no file
            is created.
            return_response (bool, optional): Set it True if you want it to
            return the actual response object.
            By default, it's False, which means a dictionary will be returned.

        Raises:
            ValueError: returns the error message from the API in case one
            (or more) of your request parameters are incorrect.

        Returns:
            Data sent back from the API. Either a response or dict object
            depending on the *return_response* argument.
        """
        response = requests.get(self.api_url, params=params)
        if response.status_code == 400:
            raise ValueError(response.text)
        elif response.status_code == 504:
            raise TimeoutError("The server reported a gateway time-out error.")
        if export_file_name != "":
            utils.export_file(response.content, export_file_name)
        if return_response:
            return response
        return response.json()


class Events(OpenseaBase):

    MAX_API_ITEMS = 300

    def __init__(self):
        """Endpoint to fetch data about multiple events.
        More info about this endpoint in the OpenSea docs:
        https://docs.opensea.io/reference/retrieving-asset-events
        """
        super().__init__(endpoint="events")

    def fetch(
        self,
        asset_contract_address=None,
        collection_slug=None,
        token_id=None,
        account_address=None,
        event_type=None,
        only_opensea=False,
        auction_type=None,
        offset=0,
        limit=None,
        occurred_before=None,
        occurred_after=None,
        export_file_name="",
    ):
        """Fetches Events data from the API. Function arguments will be passed
        as API query parameters.

        OpenSea API Events query parameters:
        https://docs.opensea.io/reference/retrieving-asset-events

        All arguments will be passed without modification, except
        *occurred_before* and *occurred_after*. For these two args, you need
        to use datetime objects when calling this function.

        There's one extra optional argument:
        export_file_name (str, optional): Exports the JSON data into a the
        specified file.

        Returns:
            [dict]: Events data
        """
        if (occurred_after is not None
            and not isinstance(occurred_after, datetime)):
            raise ValueError("occurred_after must be a datetime object")

        if (occurred_before is not None
            and not isinstance(occurred_before, datetime)):
            raise ValueError("occurred_before must be a datetime object")

        query_params = {
            "asset_contract_address": asset_contract_address,
            "collection_slug": collection_slug,
            "token_id": token_id,
            "account_address": account_address,
            "event_type": event_type,
            "only_opensea": only_opensea,
            "auction_type": auction_type,
            "offset": offset,
            "limit": self.MAX_API_ITEMS if limit is None else limit,
        }
        if occurred_before is not None:
            query_params["occurred_before"] = occurred_before.timestamp()
        if occurred_after is not None:
            query_params["occurred_after"] = occurred_after.timestamp()
        return super()._make_request(query_params, export_file_name)


class Asset(OpenseaBase):
    def __init__(self, asset_contract_address, token_id):
        """Endpoint to fetch data about a single asset.
        More info about this endpoint in the OpenSea docs:
        https://docs.opensea.io/reference/retrieving-a-single-asset

        Args:
            asset_contract_address (str)
            token_id (str):
        """
        super().__init__(endpoint=f"asset/{asset_contract_address}/{token_id}")

    def fetch(self, account_address=None, export_file_name=""):
        """Fetches Asset data from the API.

        Args:
            account_address (str, optional). Defaults to None.
            export_file_name (str, optional): Exports the JSON data into a the
            specified file.

        Returns:
            [dict]: Single asset data
        """
        query_params = {"account_address": account_address}
        return super()._make_request(query_params, export_file_name)


class Assets(OpenseaBase):

    MAX_API_ITEMS = 50

    def __init__(self):
        """Endpoint to fetch data about multiple assets.
        More info about this endpoint in the OpenSea docs:
        https://docs.opensea.io/reference/getting-assets
        """
        super().__init__(endpoint="assets")

    def fetch(
        self,
        owner=None,
        token_ids=(),
        asset_contract_address=None,
        asset_contract_addresses=None,
        order_by=None,
        order_direction=None,
        offset=None,
        limit=None,
        collection=None,
        cursor=None,
        export_file_name="",
    ):
        """Fetches assets data from the API. Function arguments will be passed
        as API query parameters, without modification.

        OpenSea API Assets query parameters:
        https://docs.opensea.io/reference/getting-assets

        There's one extra optional argument:
        export_file_name (str, optional): Exports the JSON data into a the
        specified file.

        Returns:
            [dict]: Assets data
        """
        query_params = {
            "owner": owner,
            "token_ids": token_ids,
            "asset_contract_address": asset_contract_address,
            "asset_contract_addresses": asset_contract_addresses,
            "order_by": order_by,
            "order_direction": order_direction,
            "offset": offset,
            "limit": self.MAX_API_ITEMS if limit is None else limit,
            "collection": collection,
            "cursor": cursor
        }
        return super()._make_request(query_params, export_file_name)


class Contract(OpenseaBase):
    def __init__(self, asset_contract_address):
        """Endpoint to fetch data about a single asset contract.
        More info about this endpoint in the OpenSea docs:
        https://docs.opensea.io/reference/retrieving-a-single-contract

        Args:
            asset_contract_address (str)
        """
        super().__init__(endpoint=f"asset_contract/{asset_contract_address}")

    def fetch(self, export_file_name=""):
        """Fetches asset contract data from the API.

        OpenSea API Asset Contract query parameters:
        https://docs.opensea.io/reference/retrieving-a-single-contract

        Args:
            export_file_name (str, optional): Exports the JSON data into a the
            specified file.

        Returns:
            [dict]: Single asset contract data
        """
        return super()._make_request(None, export_file_name)


class Collection(OpenseaBase):
    def __init__(self, collection_slug):
        """Endpoint to fetch data about a single asset contract.
        More info about this endpoint in the OpenSea docs:
        https://docs.opensea.io/reference/retrieving-a-single-collection

        Args:
            collection_slug (str)
        """
        super().__init__(endpoint=f"collection/{collection_slug}")

    def fetch(self, export_file_name=""):
        """Fetches collection data from the API.

        OpenSea API Collection query parameters:
        https://docs.opensea.io/reference/retrieving-a-single-collection

        Args:
            export_file_name (str, optional): Exports the JSON data into a the
            specified file.

        Returns:
            [dict]: Single collection data
        """
        return super()._make_request(None, export_file_name)


class CollectionStats(OpenseaBase):
    def __init__(self, collection_slug):
        """Endpoint to fetch a single collection's stats.
        More info about this endpoint in the OpenSea docs:
        https://docs.opensea.io/reference/retrieving-collection-stats

        Args:
            collection_slug (str)
        """
        super().__init__(endpoint=f"collection/{collection_slug}/stats")

    def fetch(self, export_file_name=""):
        """Fetches collection stats data from the API.

        OpenSea API Collection Stats query parameters:
        https://docs.opensea.io/reference/retrieving-collection-stats

        Args:
            export_file_name (str, optional): Exports the JSON data into the
            specified file.

        Returns:
            [dict]: Collection stats
        """
        return super()._make_request(None, export_file_name)


class Collections(OpenseaBase):

    MAX_API_ITEMS = 300

    def __init__(self):
        """Endpoint to fetch data about multiple collections.
        More info about this endpoint in the OpenSea docs:
        https://docs.opensea.io/reference/retrieving-collections
        """
        super().__init__(endpoint="collections")

    def fetch(self, asset_owner=None, offset=None, limit=None,
              export_file_name=""):
        """Fetches Collections data from the API. Function arguments will be
        passed as API query parameters,
        without modification.

        OpenSea API Collections query parameters:
        https://docs.opensea.io/reference/retrieving-collections

        There's one extra optional argument:
        export_file_name (str, optional): Exports the JSON data into a the
        specified file.

        Returns:
            [dict]: Collections data
        """
        query_params = {
            "asset_owner": asset_owner,
            "offset": offset,
            "limit": self.MAX_API_ITEMS if limit is None else limit,
        }
        return super()._make_request(query_params, export_file_name)


class Bundles(OpenseaBase):

    MAX_API_ITEMS = 50

    def __init__(self):
        super().__init__(endpoint="bundles")

    def fetch(
        self,
        on_sale=None,
        owner=None,
        asset_contract_address=None,
        asset_contract_addresses=[],
        token_ids=[],
        limit=None,
        export_file_name="",
        offset=None,
    ):
        """Fetches Bundles data from the API. Function arguments will be
        passed as API query parameters,
        without modification.

        OpenSea API Bundles query parameters:
        https://docs.opensea.io/reference/retrieving-bundles

        There's one extra optional argument:
        export_file_name (str, optional): Exports the JSON data into a the
        specified file.

        Returns:
            [dict]: Bundles data
        """
        query_params = {
            "on_sale": on_sale,
            "owner": owner,
            "asset_contract_address": asset_contract_address,
            "asset_contract_addresses": asset_contract_addresses,
            "token_ids": token_ids,
            "limit": self.MAX_API_ITEMS if limit is None else limit,
            "offset": offset,
        }
        return super()._make_request(query_params, export_file_name)


opensea/opensea_api.py
import time
import requests
from datetime import datetime
from opensea import utils


class OpenseaAPI:

    MAX_EVENT_ITEMS = 300
    MAX_ASSET_ITEMS = 50
    MAX_COLLECTION_ITEMS = 300
    MAX_BUNDLE_ITEMS = 50
    MAX_LISTING_ITEMS = 50
    MAX_OFFER_ITEMS = 50

    def __init__(self, base_url="https://api.opensea.io/api", apikey=None,
                 version="v1"):
        """Base class to interact with the OpenSea API and fetch NFT data.

        Args:
            base_url (str): OpenSea API base URL. Defaults to
            "https://api.opensea.io/api".
            apikey (str): OpenSea API key (you need to request one)
            version (str, optional): API version. Defaults to "v1".
        """
        self.api_url = f"{base_url}/{version}"
        self.apikey = apikey

    def _make_request(self, endpoint=None, params=None, export_file_name="",
                      return_response=False):
        """Makes a request to the OpenSea API and returns either a response
        object or dictionary.

        Args:
            endpoint (str, optional): API endpoint to use for the request.
            params (dict, optional): Query parameters to include in the
            request. Defaults to None.
            export_file_name (str, optional): In case you want to download the
            data into a file,
            specify the filename here. Eg. 'export.json'. Be default, no file
            is created.
            return_response (bool, optional): Set it True if you want it to
            return the actual response object.
            By default, it's False, which means a dictionary will be returned.
            next_url (str, optional): If you want to paginate, provide the
            `next` value here (this is a URL) OpenSea provides in the response.
            If this argument is provided, `endpoint` will be ignored.

        Raises:
            ValueError: returns the error message from the API in case one
            (or more) of your request parameters are incorrect.
            HTTPError: Unauthorized access. Try using an API key.
            ConnectionError: your request got blocked by the server (try
            using an API key if you keep getting this error)
            TimeoutError: your request timed out (try rate limiting)

        Returns:
            Data sent back from the API. Either a response or dict object
            depending on the `return_response` argument.
        """
        if endpoint is None:
            raise ValueError(
                """You need to define an `endpoint` when
                             making a request!"""
            )

        headers = {"X-API-KEY": self.apikey}
        url = f"{self.api_url}/{endpoint}"
        response = requests.get(url, headers=headers, params=params)
        if response.status_code == 400:
            raise ValueError(response.text)
        elif response.status_code == 401:
            raise requests.exceptions.HTTPError(response.text)
        elif response.status_code == 403:
            raise ConnectionError("The server blocked access.")
        elif response.status_code == 495:
            raise requests.exceptions.SSLError("SSL certificate error")
        elif response.status_code == 504:
            raise TimeoutError("The server reported a gateway time-out error.")

        if export_file_name != "":
            utils.export_file(response.content, export_file_name)
        if return_response:
            return response
        return response.json()

    def events(
        self,
        asset_contract_address=None,
        collection_slug=None,
        token_id=None,
        account_address=None,
        event_type=None,
        only_opensea=False,
        auction_type=None,
        limit=None,
        occurred_before=None,
        occurred_after=None,
        collection_editor=None,
        export_file_name="",
    ):
        """Fetches Events data from the API. Function arguments will be passed
        as API query parameters.

        OpenSea API Events query parameters:
        https://docs.opensea.io/reference/retrieving-asset-events

        All arguments will be passed without modification, except
        *occurred_before* and *occurred_after*. For these two args, you need
        to use datetime objects when calling this function.

        There's one extra optional argument:
        export_file_name (str, optional): Exports the JSON data into the
        specified file.

        Returns:
            [dict]: Events data
        """
        if occurred_before is not None and not isinstance(occurred_before,
                                                          datetime):
            raise ValueError("`occurred_before` must be a datetime object")

        if occurred_after is not None and not isinstance(occurred_after,
                                                         datetime):
            raise ValueError("`occurred_after` must be a datetime object")

        query_params = {
            "asset_contract_address": asset_contract_address,
            "collection_slug": collection_slug,
            "token_id": token_id,
            "account_address": account_address,
            "event_type": event_type,
            "only_opensea": only_opensea,
            "auction_type": auction_type,
            "collection_editor": collection_editor,
            "limit": self.MAX_EVENT_ITEMS if limit is None else limit,
        }
        if occurred_before is not None:
            query_params["occurred_before"] = occurred_before.timestamp()

        if occurred_after is not None:
            query_params["occurred_after"] = occurred_after.timestamp()
        return self._make_request("events", query_params, export_file_name)

    def events_backfill(
        self,
        start,
        until,
        rate_limiting=2,
        asset_contract_address=None,
        collection_slug=None,
        token_id=None,
        account_address=None,
        event_type=None,
        only_opensea=False,
        auction_type=None,
        limit=None,
        collection_editor=None,
    ):
        """
        EXPERIMENTAL FUNCTION!

        Expected behaviour:

        Download events and paginate over multiple pages until the given
        time is reached. Pagination happens **backwards** (so you can use this
        function to **backfill** events data eg. into a database) from `start`
        until `until`.

        The function returns a generator.

        Args:
            start (datetime): A point in time where you want to start
            downloading data from.
            until (datetime): How much data do you want?
            How much do you want to go back in time? This datetime value will
            provide that threshold.
            rate_limiting (int, optional): Seconds to wait between requests.
            Defaults to 2.

            Other parameters are available (all of the `events` endpoint
            parameters) and they are documented in the OpenSea docs
            https://docs.opensea.io/reference/retrieving-asset-events

        Yields:
            dictionary: event data
        """
        if not isinstance(until, datetime) or not isinstance(start, datetime):
            raise ValueError("`until` and `start` must be datetime objects")

        if until > start:
            raise ValueError(
                """`start` must be a later point in time
                             than `until`"""
            )

        query_params = {
            "asset_contract_address": asset_contract_address,
            "collection_slug": collection_slug,
            "token_id": token_id,
            "account_address": account_address,
            "event_type": event_type,
            "only_opensea": only_opensea,
            "auction_type": auction_type,
            "limit": self.MAX_EVENT_ITEMS if limit is None else limit,
            "occurred_before": start,
            "collection_editor": collection_editor,
        }

        # make the first request to get the `next` cursor value
        first_request = self._make_request("events", query_params)
        yield first_request
        query_params["cursor"] = first_request["next"]

        # paginate
        while True:
            time.sleep(rate_limiting)
            data = self._make_request("events", query_params)

            # update the `next` parameter for the upcoming request
            query_params["cursor"] = data["next"]

            time_field = data["asset_events"][0]["created_date"]
            current_time = utils.str_to_datetime_utc(time_field)
            if current_time >= until:
                yield data
            else:
                break
        yield None

    def asset(
        self,
        asset_contract_address,
        token_id,
        account_address=None,
        include_orders=False,
        export_file_name="",
    ):
        """Fetches Asset data from the API.

        Args:
            OpenSea API Asset query parameters:
            https://docs.opensea.io/reference/retrieving-a-single-asset

            Extra args:
            export_file_name (str, optional): Exports the JSON data into a the
            specified file.

        Returns:
            [dict]: Single asset data
        """
        endpoint = f"asset/{asset_contract_address}/{token_id}"
        query_params = {"account_address": account_address,
                        "include_orders": include_orders}
        return self._make_request(endpoint, query_params, export_file_name)

    def assets(
        self,
        owner=None,
        token_ids=[],
        asset_contract_address=None,
        asset_contract_addresses=None,
        order_by=None,
        order_direction=None,
        offset=None,
        limit=None,
        collection=None,
        include_orders=False,
        pagination=False,
        rate_limiting=2,
        export_file_name="",
    ):
        """Fetches assets data from the API. Function arguments will be passed
        as API query parameters, without modification.

        OpenSea API Assets query parameters:
        https://docs.opensea.io/reference/getting-assets

        There are extra optional arguments:
        pagination (boolean, optional): Whether you want to get only the first
        page of assets, or all of them. If it's `True` it will use the
        cursor-based pagination provided by OpenSea. Defaults to False.
        rate_limiting (int, optional): Only relevant if pagination is `True`.
        It applies a rate limitation in-between requests. Defaults to 2 sec.
        export_file_name (str, optional): Exports the JSON data into the
        specified file. If pagination is `True` this argument is ignored.

        Returns:
            [dict]: Assets data
        """
        query_params = {
            "owner": owner,
            "token_ids": token_ids,
            "asset_contract_address": asset_contract_address,
            "asset_contract_addresses": asset_contract_addresses,
            "order_by": order_by,
            "order_direction": order_direction,
            "offset": offset,
            "limit": self.MAX_ASSET_ITEMS if limit is None else limit,
            "collection": collection,
            "include_orders": include_orders
        }
        if pagination:
            # make the first request to get the `next` cursor value
            first_request = self._make_request("assets", query_params)
            yield first_request
            query_params["cursor"] = first_request.get("next")

            # paginate
            while True:
                time.sleep(rate_limiting)
                if query_params["cursor"] is not None:
                    response = self._make_request("assets", query_params)
                    yield response
                    query_params["cursor"] = response.get("next")
                else:
                    break  # stop pagination if there is no next page
        else:
            return self._make_request("assets", query_params, export_file_name)

    def contract(self, asset_contract_address, export_file_name=""):
        """Fetches asset contract data from the API.

        OpenSea API Asset Contract query parameters:
        https://docs.opensea.io/reference/retrieving-a-single-contract

        Args:
            asset_contract_address (str): Contract address of the NFT.
            export_file_name (str, optional): Exports the JSON data into a the
            specified file.

        Returns:
            [dict]: Single asset contract data
        """
        endpoint = f"asset_contract/{asset_contract_address}"
        return self._make_request(endpoint, export_file_name=export_file_name)

    def collection(self, collection_slug, export_file_name=""):
        """Fetches collection data from the API.

        OpenSea API Collection query parameters:
        https://docs.opensea.io/reference/retrieving-a-single-collection

        Args:
            collection_slug (str): Collection slug (unique identifer)
            export_file_name (str, optional): Exports the JSON data into a the
            specified file.

        Returns:
            [dict]: Single collection data
        """
        endpoint = f"collection/{collection_slug}"
        return self._make_request(endpoint, export_file_name=export_file_name)

    def collection_stats(self, collection_slug, export_file_name=""):
        """Fetches collection stats data from the API.

        OpenSea API Collection Stats query parameters:
        https://docs.opensea.io/reference/retrieving-collection-stats

        Args:
            export_file_name (str, optional): Exports the JSON data into the
            specified file.

        Returns:
            [dict]: Collection stats
        """
        endpoint = f"collection/{collection_slug}/stats"
        return self._make_request(endpoint, export_file_name=export_file_name)

    def collections(
        self, asset_owner=None, offset=None, limit=None, export_file_name=""
    ):
        """Fetches Collections data from the API. Function arguments will be
        passed as API query parameters,
        without modification.

        OpenSea API Collections query parameters:
        https://docs.opensea.io/reference/retrieving-collections

        There's one extra optional argument:
        export_file_name (str, optional): Exports the JSON data into a the
        specified file.

        Returns:
            [dict]: Collections data
        """
        query_params = {
            "asset_owner": asset_owner,
            "offset": offset,
            "limit": self.MAX_COLLECTION_ITEMS if limit is None else limit,
        }
        return self._make_request("collections", query_params,
                                  export_file_name)

    def bundles(
        self,
        on_sale=None,
        owner=None,
        asset_contract_address=None,
        asset_contract_addresses=[],
        token_ids=[],
        limit=None,
        export_file_name="",
        offset=None,
    ):
        """Fetches Bundles data from the API. Function arguments will be
        passed as API query parameters,
        without modification.

        OpenSea API Bundles query parameters:
        https://docs.opensea.io/reference/retrieving-bundles

        There's one extra optional argument:
        export_file_name (str, optional): Exports the JSON data into a the
        specified file.

        Returns:
            [dict]: Bundles data
        """
        query_params = {
            "on_sale": on_sale,
            "owner": owner,
            "asset_contract_address": asset_contract_address,
            "asset_contract_addresses": asset_contract_addresses,
            "token_ids": token_ids,
            "limit": self.MAX_BUNDLE_ITEMS if limit is None else limit,
            "offset": offset,
        }
        return self._make_request("bundles", query_params, export_file_name)

    def listings(
        self, asset_contract_address, token_id, limit=None, export_file_name=""
    ):
        """Fetches Listings data for an asset from the API. Function arguments
        will be passed as API query parameters, without modification.

        OpenSea API Listings query parameters:
        https://docs.opensea.io/reference/asset-listings

        There's one extra optional argument:
        export_file_name (str, optional): Exports the JSON data into a the
        specified file.

        Returns:
            [dict]: Listings data
        """
        query_params = {
            "limit": self.MAX_LISTING_ITEMS if limit is None else limit
        }
        endpoint = f"asset/{asset_contract_address}/{token_id}/listings"
        return self._make_request(endpoint, query_params, export_file_name)

    def offers(self, asset_contract_address, token_id, limit=None,
               export_file_name=""):
        """Fetches Offers data for an asset from the API. Function arguments
        will be passed as API query parameters, without modification.

        OpenSea API Listings query parameters:
        https://docs.opensea.io/reference/asset-offers

        There's one extra optional argument:
        export_file_name (str, optional): Exports the JSON data into a the
        specified file.

        Returns:
            [dict]: Offers data
        """
        query_params = {
            "limit": self.MAX_OFFER_ITEMS if limit is None else limit
        }
        endpoint = f"asset/{asset_contract_address}/{token_id}/offers"
        return self._make_request(endpoint, query_params, export_file_name)
